### ISSUE TITLE
Overriding SIGINT doesn't work as expected in the new REPL

### ISSUE THREAD

**Issue by pablogsal** [python/cpython#120221]
# Bug report

In the old REPL, overriding SIGINT and pressing Ctrl-C works as expected:

```
❯ python
Python 3.11.6 (main, Nov  3 2023, 17:05:41) [Clang 15.0.0 (clang-1500.0.40.1)] on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>> import signal
>>> signal.signal(signal.SIGINT, lambda *x: print("NOOOO"))
<built-in function default_int_handler>
>>> NOOOO
NOOOO
```

but in the new REPL it doesn't trigger the signal handler:

```
❯ ./python.exe
Python 3.14.0a0 (heads/more_offsets-dirty:9403c2cf58, Jun  6 2024, 12:10:11) [Clang 15.0.0 (clang-1500.3.9.4)] on darwin
Type "help", "copyright", "credits" or "license" for more information.
>>> import signal
>>> signal.signal(signal.SIGINT, lambda *x: print("NOOOO"))
<built-in function default_int_handler>
>>>
KeyboardInterrupt
>>>
```

### Linked PRs
* gh-120354
* gh-120368
* gh-123795
* gh-123799

**Comment by pablogsal** [comment]
CC: @lysnikolaou @ambv

**Comment by pablogsal** [comment]
This will be fixed by https://github.com/python/cpython/pull/120234

**Comment by vstinner** [comment]
I noticed the problem when trying to debug Python in gdb. CTRL+C doesn't go into gdb anymore. For now, I'm using a bash alias:

```
alias gdb='PYTHON_BASIC_REPL=1 gdb'
```

**Comment by vstinner** [comment]
Thank you. CTRL+C is gdb stops again the process as expected :)

**Comment by eryksun** [comment]
This could have been addressed by changing `default_keymap` to map `(r"\x03", "interrupt")` instead of mapping it to the `ctrl_c` command. The `interrupt` command calls `os.kill(os.getpid(), signal.SIGINT)`, which works fine for overriding the `SIGINT` handler and supporting gdb. That said, the decision to let the terminal send the signal makes it equivalent to how it's currently implemented on Windows, and consistency is a plus.

**Comment by pablogsal** [comment]
> This could have been addressed by changing `default_keymap` in "Lib/_pyepl/reader.py" to map `(r"\x03", "interrupt")` instead of mapping it to the `ctrl_c` command. The `interrupt` command calls `os.kill(os.getpid(), signal.SIGINT)`, which works fine for overriding the `SIGINT` handler and supporting gdb. That said, the decision to let the terminal send the signal makes it equivalent to how it's currently implemented on Windows, and consistency is a plus.

The problem of that approach is that if you paste a huge buffer and then want to cancel because pasting it's too slow then you need to wait until you get the key press for Ctrl-C to send the signal while using signals directly allows us to cancel almost immediately

**Comment by ambv** [comment]
This broke CTRL-C in the asyncio REPL. Now the `interrupt` and `ctrl_c` commands are never executed. I need some more time to investigate what happens on Windows to not break that when fixing Linux/Mac.

**Comment by ambv** [comment]
Fixed by #123799.
